rfc-drafts/stream.md

```diff
@@ -0,0 +1,186 @@
+  task to be re-awoken once data is ready.
+
+[`Future::poll`]: https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll
+[pinned]: https://doc.rust-lang.org/std/pin/struct.Pin.html
+[context]: https://doc.rust-lang.org/std/task/struct.Context.html
+[Waker]: https://doc.rust-lang.org/std/task/struct.Waker.html
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This section goes into details about various aspects of the design and
+why they ended up the way they did.
+
+## Why use a `poll` method?
+
+An alternative design for the stream trait would be to have a trait
+that defines an async `next` method:
+
+```rust
+trait Stream {
+    type Item;
+    
+    async fn next(&mut self) -> Option<Self::Item>;
+}
+```
+
+Unfortunately, async methods in traits are not currently supported,
+and there [are a number of challenges to be
+resolved](https://rust-lang.github.io/wg-async-foundations/design_notes/async_fn_in_traits.html)
+before they can be added. 
+
+Moreover, it is not clear yet how to make traits that contain async
+functions be `dyn` safe, and it is imporant to be able to pass around `dyn
+Stream` values without the need to monomorphize the functions that work
+with them.
+
+Unfortunately, the use of poll does mean that it is harder to write
+stream implementations.
+
+## What about combinators?
+
+The `Iterator` trait defines a number of useful combinators, like
+`map`.  The `Stream` trait being proposed here does not include any
+such conveniences.  Instead, they are available via extension traits,
+such as the [`StreamExt`] trait offered by the [`futures`] crate.
+
+[`StreamExt`]: https://docs.rs/futures/0.3.5/futures/stream/trait.StreamExt.html
+[`futures`]: https://crates.io/crates/futures
+
+The reason that we have chosen to exclude combinators is that a number
+of them would require access to async closures. As of this writing,
+async closures are unstable and there are a number of [outstanding
+design issues] to be resolved before they are added. Therefore, we've
+decided to enable progress on the stream trait by stabilizing a core,
+and to come back to the problem of extending it with combinators.
+
+[outstanding design issues]: https://rust-lang.github.io/wg-async-foundations/design_notes/async_closures.html
+
+This path does carry some risk. Adding combinator methods can cause
+existing code to stop compiling due to the ambiguities in method
+resolution. We have had problems in the past with attempting to migate
+iterator helper methods from `itertools` for this same reason.
+
+While such breakage is technically permitted by our semver guidelines,
+it would obviously be best to avoid it, or at least to go to great
+lengths to mitigate its effects. One option would be to extend the
+language to allow method resolution to "favor" the extension trait in
+existing code, perhaps as part of an edition migration.
+
+## "Attached" streams
```

Comment:
It'd be great if this RFC could also mention async iteration syntax at least somewhere. I believe it's one of the requirements for "attached streams" to work so maybe there. Or perhaps it's noteworthy enough to mention under "future possibilities" by itself. It seems important to me to at least mention when discussing streams (: